File: iOS/Prosary/Models/MarianAntiphonOption.swift
//
//  MarianAntiphonOption.swift
//  Prosary
//

import Foundation

/// Which closing Marian antiphon (if any) follows the Rosary.
enum MarianAntiphonOption: String, Codable, CaseIterable, Identifiable {
  case none
  /// Pick the antiphon proper to the current liturgical season automatically.
  case seasonal
  case salveRegina
  case almaRedemptorisMater
  case aveReginaCaelorum
  case reginaCaeli
  case subTuumPraesidium

  var id: String { rawValue }

  var displayName: String {
    displayName(languageCode: LanguageCatalog.resolve(nil).code)
  }

  func displayName(languageCode: String) -> String {
    let name = switch self {
    case .none:                 String(localized: "marianAntiphon.none", defaultValue: "None")
    case .seasonal:             String(localized: "marianAntiphon.seasonal", defaultValue: "Automatic (Seasonal)")
    case .salveRegina:          PrayerTranslations.get(languageCode: languageCode, key: .salveReginaTitle)
    case .almaRedemptorisMater: PrayerTranslations.get(languageCode: languageCode, key: .almaRedemptorisMaterTitle)
    case .aveReginaCaelorum:    PrayerTranslations.get(languageCode: languageCode, key: .aveReginaCaelorumTitle)
    case .reginaCaeli:          PrayerTranslations.get(languageCode: languageCode, key: .reginaCaeliTitle)
    case .subTuumPraesidium:    PrayerTranslations.get(languageCode: languageCode, key: .subTuumPraesidiumTitle)
    }
    return HebrewDisplayText.unpointed(name)
  }
}

File: Android/app/src/main/java/com/dkaluta/prosary/models/MarianAntiphonOption.kt
package com.dkaluta.prosary.models

import androidx.annotation.StringRes
import com.dkaluta.prosary.R
import android.content.Context
import com.dkaluta.prosary.content.PrayerKey
import com.dkaluta.prosary.content.PrayerTranslations

/** Which closing Marian antiphon (if any) follows the Rosary. */
enum class MarianAntiphonOption {
    None,

    /** Pick the antiphon proper to the current liturgical season automatically. */
    Seasonal,
    SalveRegina,
    AlmaRedemptorisMater,
    AveReginaCaelorum,
    ReginaCaeli,
    SubTuumPraesidium;

    @get:StringRes
    val displayNameRes: Int
        get() = when (this) {
            None -> R.string.antiphon_none
            Seasonal -> R.string.antiphon_seasonal
            SalveRegina -> R.string.antiphon_salve_regina
            AlmaRedemptorisMater -> R.string.antiphon_alma_redemptoris
            AveReginaCaelorum -> R.string.antiphon_ave_regina
            ReginaCaeli -> R.string.antiphon_regina_caeli
            SubTuumPraesidium -> R.string.antiphon_sub_tuum
        }

    fun displayName(context: Context, languageCode: String): String = when (this) {
        None, Seasonal -> context.getString(displayNameRes)
        SalveRegina -> PrayerTranslations.get(languageCode, PrayerKey.SalveReginaTitle)
        AlmaRedemptorisMater -> PrayerTranslations.get(languageCode, PrayerKey.AlmaRedemptorisMaterTitle)
        AveReginaCaelorum -> PrayerTranslations.get(languageCode, PrayerKey.AveReginaCaelorumTitle)
        ReginaCaeli -> PrayerTranslations.get(languageCode, PrayerKey.ReginaCaeliTitle)
        SubTuumPraesidium -> PrayerTranslations.get(languageCode, PrayerKey.SubTuumPraesidiumTitle)
    }
}
